File: 735-asteroid-collision/735-asteroid-collision.scala
import scala.collection.mutable._

object Solution {
    def asteroidCollision(asteroids: Array[Int]): Array[Int] = {
        val stack: Stack[Int] = Stack()

        asteroids.foreach { a =>
            if (a > 0) {
                stack.push(a)
            } else {
                while(stack.nonEmpty && stack.top > 0 && stack.top + a < 0) {
                    stack.pop()
                }
                if (stack.isEmpty || stack.top < 0) stack.push(a)
                if (stack.top == -a) stack.pop()
            }
        }

        stack.toArray.reverse
    }
}

File: 735-asteroid-collision/735-asteroid-collision.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Deque<Integer> stack = new LinkedList<>();
        
        for(int a : asteroids) {
            if (a > 0) {
                stack.push(a);
                continue;
            }
            
            while(!stack.isEmpty() && stack.peek() > 0 && stack.peek() < Math.abs(a)) {
                stack.pop();
            }
            
            if (stack.isEmpty() || stack.peek() < 0) stack.push(a); 
            else if (!stack.isEmpty() && stack.peek() == -a) stack.pop();
        }
        
        
        int[] res = new int[stack.size()];
        int idx = 0;
        
        while(!stack.isEmpty()) {
            res[idx++] = stack.pollLast();
        }
        
        return res;
    }
}